analyzer/src/funTest/assets/projects/synthetic/carthage-expected-output.yml

```diff
@@ -0,0 +1,123 @@
+---
+project:
+  id: "Carthage:oss-review-toolkit:ort:"
+  definition_file_path: "<REPLACE_DEFINITION_FILE_PATH>"
+  declared_licenses: []
+  declared_licenses_processed: {}
+  vcs:
+    type: ""
+    url: ""
+    revision: ""
+    path: ""
+  vcs_processed:
+    type: "Git"
+    url: "<REPLACE_URL>"
+    revision: "<REPLACE_REVISION>"
+    path: "<REPLACE_PATH>"
+  homepage_url: ""
+  scopes: []
+packages:
+- id: "Carthage::binaryspec:1.0.0"
+  purl: "pkg:carthage/binaryspec@1.0.0"
+  declared_licenses: []
+  declared_licenses_processed: {}
+  description: ""
+  homepage_url: ""
+  binary_artifact:
+    url: "https://path/to/binary/artifact.zip"
+    hash:
+      value: ""
+      algorithm: ""
+  source_artifact:
+    url: "https://path/to/binary/artifact.zip"
+    hash:
+      value: ""
+      algorithm: ""
+  vcs:
+    type: ""
+    url: ""
+    revision: ""
+    path: ""
+  vcs_processed:
+    type: ""
+    url: ""
+    revision: ""
+    path: ""
+- id: "Carthage:Alamofire:Alamofire:5.2.2"
+  purl: "pkg:carthage/Alamofire/Alamofire@5.2.2"
+  declared_licenses: []
+  declared_licenses_processed: {}
+  description: ""
+  homepage_url: "https://github.com/Alamofire/Alamofire"
+  binary_artifact:
+    url: ""
+    hash:
+      value: ""
+      algorithm: ""
+  source_artifact:
+    url: "https://api.github.com/repos/Alamofire/Alamofire/tarball/5.2.2"
+    hash:
+      value: ""
+      algorithm: ""
+  vcs:
+    type: "Git"
+    url: "https://github.com/Alamofire/Alamofire.git"
+    revision: "5.2.2"
+    path: ""
+  vcs_processed:
+    type: "Git"
+    url: "https://github.com/Alamofire/Alamofire.git"
+    revision: "5.2.2"
+    path: ""
+- id: "Carthage:ReactiveCocoa:ReactiveCocoa:11.0.0"
+  purl: "pkg:carthage/ReactiveCocoa/ReactiveCocoa@11.0.0"
+  declared_licenses: []
+  declared_licenses_processed: {}
+  description: ""
+  homepage_url: "https://github.com/ReactiveCocoa/ReactiveCocoa"
+  binary_artifact:
+    url: ""
+    hash:
+      value: ""
+      algorithm: ""
+  source_artifact:
+    url: "https://api.github.com/repos/ReactiveCocoa/ReactiveCocoa/tarball/11.0.0"
```

Comment:
@mnonnenmacher, @sschuberth what do you think?

Reply:
I agree it makes sense to set both, VcsInfo and source artifact, if we can clearly deduce the VcsInfo. IMO it's the responsibility of the downloader / scanner which of the two to take then. We already have the same situation e.g. for Maven artifacts which have both a sources artifact published and a valid `<scm>` tag in the POM. Currently we [hard-code to always prefer VCS downloads](https://github.com/oss-review-toolkit/ort/blob/621830f67a420519b9e1808731668dcab5105f28/downloader/src/main/kotlin/Downloader.kt#L125-L156) over source artifacts, but there have been requests in the past to change this / make this configurable.

Reply:
@sschuberth - it seems I wasn't clear enough: the reason I raised this here was because in this case the content of the source artifact always equals the content of the VCS location (I hope this assumption of mine is correct?). This is different compared to Maven artifacts and thus IMO a different argumentation may apply.

Note: For the downloader / scanner there is no way of knowing that the contents do equal. This knowledge is only present in this package manager.

Reply:
Yes, I got that difference, but I believe the same argumentation applies.

Reply:
I actually see only cons of this approach, but no benefits
- maintainance: A bunch of VCS host specific code is added
- redundant scans: If the `scanner` is made configurable which location to scan, including the option to scan both locations, then redundant scan are made. This degrades execution time and increases storage space needs.

In short, I still prefer as a general rule to only set one location, either VCS or source artifact, in case we know that the contents do actually equal.

So it seems we disagree. Let's see what other opinions we have...

Reply:
> maintainance: A bunch of VCS host specific code is added

Why? Where? All VCS host specific functionality should be / is already part of the VcsHost class, so parsing VcsInfo from given URLs should add only very little code to this package manager.

> If the scanner is made configurable which location to scan, including the option to scan both locations, then redundant scans are made.

*If* the scanner would be allowed to scan both / all sources, then you'd be right. However, at least I was never considering this to be a configuration option. Esp. #2848 is about being able to configure the *order* of preference, but still only one source would ever be scanned. And I currently cannot think of a case where it would make sense to scan multiple source locations for a single package.

> In short, I still prefer as a general rule to only set one location, either VCS or source artifact, in case we know that the contents do actually equal.

I'm mostly fine with either way. But if we only keep one location, it should be the VCS location in this case. Also, I'd make your general rule a bit more strict: We should only set one / the VCS location if we know it to be equal to the source artifact *and* the source artiafact URL can be easily deduced from the VCS information (i.e. is available from the same host), like in this case.

Reply:
> Why? Where? All VCS host specific functionality should be / is already part of the VcsHost class, so parsing VcsInfo from given URLs should add only very little code to this package manager.

Because this PR is adding code to the VcsHost class and its tests. 

> at least I was never considering this to be a configuration option. 

I remembered the opposite, interesting that it's not the case.
If scanning both is definetely not going to happen the issue becomes less critical.

...but do you actually agree with that setting the source artifact (in addition to VCS) adds complexity to the code?

Reply:
> I remembered the opposite, interesting that it's not the case.

Then this must have been a misunderstanding.

> ...but do you actually agree with that setting the source artifact (in addition to VCS) adds complexity to the code?

Sure it does add some code, but it's debatable how "complex" that code is. Also, I'm not really counting the new VcsHost code towards the complexity of the Carthage implementation. That's a bit unfair as the VcsHost code is generic and could be used in more places in the future, too.

Reply:
> Sure it does add some code, but it's debatable how "complex" that code is. Also, I'm not really counting the new VcsHost code towards the complexity of the Carthage implementation. That's a bit unfair as the VcsHost code is generic and could be used in more places in the future, too.

Do you foresee any future use cases where its used for something different than setting the source artifact URL to point to a location with exactly the same content as the VCS?

Reply:
I don't like that we set the source artifact to API of GitHub, GitLab or BitBucket for several reasons:
- Carthage does not seem to have the concept of sources artifact so why are we artificially creating one? Simply set the VCS, if users want to use RestFul API to download from VCS sources then this should be a configurable option in the Downloader but not something it does by itself.
- Introducing reliance on Restful APIs makes it difficult to run ORT in fully offline mode
- Restful APIs are often rate limited e.g. GitHub only allow 5000 calls per hour whilst clones are not limited
- Restful APIs change more often then CLIs

Reply:
Thanks for the clarification.
I removed the `sourceArtifact`s for dependencies from git / github.